### ISSUE TITLE
add clutter identification to layer selection and averaging

### ISSUE THREAD

**Issue by adokter** [jspaaks/baltrad-node#12]
The function `classifyGates` classifies which gates should be included in the altitude layer averages.

`svdfit` and the layer averaging (of reflectivity etc) does not take place on the same subset of gates: the set of gates fed to `svdfit` is considerably larger.

You might want to port the functionality of `classifyGates` to the function `getListOfSelectedGates`, where this function gets an additional argument that indicates whether it is selecting for `svdfit` (as set up now), or selecting for reflectivity averaging (functionality still missing and now present in `classifyGates`). For the latter, the following additional selections need to be made, depending on the profile type `iData`:

*****************************
if `iData==0` (birds-only profile) the following gates are rejected in the reflectivity averaging:
1) gates in the calculated clutter map, including fringe
2) reflectivity values higher than the threshold `dbZx` are dropped from the bird reflectivity average, because too high reflectivity values cannot be due to birds, see:
https://github.com/NLeSC/enram/blob/611297849897019eaefb3111f819c5105ba3fd2c/ncradar/lib/vol2bird/libvol2bird.c#L529
3) reflectivity without radial velocity information is discarded:
https://github.com/NLeSC/enram/blob/611297849897019eaefb3111f819c5105ba3fd2c/ncradar/lib/vol2bird/libvol2bird.c#L502-L508
4) gates in the static clutter map (as obtained from a separate volume file)
https://github.com/NLeSC/enram/blob/611297849897019eaefb3111f819c5105ba3fd2c/ncradar/lib/vol2bird/libvol2bird.c#L475
5) all gates with a radial velocity near zero (because these are likely clutter):
https://github.com/NLeSC/enram/blob/611297849897019eaefb3111f819c5105ba3fd2c/ncradar/lib/vol2bird/libvol2bird.c#L520

*****************************
if `iData==1` (non-birds-only profile) the following gates are rejected in the reflectivity averaging:
1) gates in the calculated clutter map, excluding fringe
2) reflectivity without radial velocity information is discarded:
https://github.com/NLeSC/enram/blob/611297849897019eaefb3111f819c5105ba3fd2c/ncradar/lib/vol2bird/libvol2bird.c#L502-L508
3) all gates with a radial velocity near zero (because these are likely clutter):
https://github.com/NLeSC/enram/blob/611297849897019eaefb3111f819c5105ba3fd2c/ncradar/lib/vol2bird/libvol2bird.c#L520

*****************************
if `iData==3` (full profile) the following gates are rejected in the reflectivity averaging:
1) reflectivity without radial velocity information is discarded:
https://github.com/NLeSC/enram/blob/611297849897019eaefb3111f819c5105ba3fd2c/ncradar/lib/vol2bird/libvol2bird.c#L502-L508
2) all gates with a radial velocity near zero (because these are likely clutter):
https://github.com/NLeSC/enram/blob/611297849897019eaefb3111f819c5105ba3fd2c/ncradar/lib/vol2bird/libvol2bird.c#L520

**Comment by adokter** [comment]
Thinking of it in the morning: there are also reasons to not port the functionality of classifyGates to the function `getListOfSelectedGates`, as the two functions intend to different things. `getListOfSelectedGates` makes a list that is used later on, while  `classifyGates` simply performs averaging without the need to store the selection in a list. So two different functions is maybe not such a bad idea after all.

**Comment by jspaaks** [comment]
I think I should rewrite and simplify ``classifyGates`` as follows. As far as I can tell, we have the following 6 conditions:

1. gates in the static clutter map
1. calculated cluttermap (without fringe)
1. fringe
1. reflectivity values higher than the threshold dbZx are dropped from the bird reflectivity average, because too high reflectivity values cannot be due to birds
1. reflectivity without radial velocity information is discarded
1. all gates with a radial velocity near zero (because these are likely clutter)

Furthermore, these conditions can all be derived from the information that is currently stored in ``points``, of which this is an extract:
```

iPoint  azim    elev    dbz         vrad        cell
       0    6.50    0.30        1.82        1.50       1
       1    7.50    0.30        2.70        5.50       1
       2    8.50    0.30        1.87        8.50       1
       3   11.50    0.30       -3.30       37.00       2
       4   12.50    0.30        1.82       31.00       2
       5   17.50    0.30        2.00       -3.00      -1
       6   19.50    0.30        3.17       25.00      -1
       7   20.50    0.30       -3.30       28.00      -1
       8   21.50    0.30        2.60       23.50      -1
       9   25.50    0.30       -3.30       -5.50      -1
      10   28.50    0.30        2.23       20.00      -1
      11   29.50    0.30        2.49       17.00      -1
      12   30.50    0.30       -3.30       19.00      -1
      13   31.50    0.30       -2.91        4.00      -1
      14   34.50    0.30        2.49      -17.00      -1
      15   35.50    0.30       -3.30      -31.50      -1
      16   37.50    0.30       -3.09       10.50      -1
      17   38.50    0.30        1.25       10.00      -1
      18   39.50    0.30        2.21        6.00      -1
      19   40.50    0.30       -3.04       -9.00      -1
      20   44.50    0.30       -3.30      -31.50      -1
      21   45.50    0.30       -1.25      -15.50      -1
```
I would like to summarize the  information on the gate by performing the 6 tests above within a new version of classifyGates, which then returns an integer code using the flagfield method as used in embedded programs: [http://en.wikipedia.org/wiki/Flag_field](http://en.wikipedia.org/wiki/Flag_field). Note that you can do this once without needing to know the type of profile it is for. 

Later on in the program, specifically, within this block: https://github.com/jspaaks/baltrad-node/blob/bd5366570952634f8927d1011fbef1170badb6a0/crave/testbaltrad.c#L449-L465, you can have a (small) function that decides whether to include the ``[azimuth, elevation]`` combination in the ``pointsSelection`` list that is passed on to ``svdfit`` based only on ``iProfileType`` and the gate code. Furthermore, this would obfuscate the need for a few of the ``nPoints<something>`` variables; instead, ``nPointsCopied`` would then contain this information. Come to think of it, maybe I should rename ``nPointsCopied`` to ``nPointsValid``, ``nPointsIncluded`` or something like that.

**Comment by adokter** [comment]
Hi Jurriaan, adding a flagfield column is a clever solution, sounds perfect!

**Comment by jspaaks** [comment]
In the table above, vrad and dbz columns have been switched by mistake

**Comment by jspaaks** [comment]
The approach proposed above has been implemented since commit 5af7b9f803b851df7578c5128f679bac619974b5 and seems to work pretty neatly. Closing the issue
